rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,308 @@
+    * remove duplication
+* Remove the routes from the candidate
+* [Retire the candidate] once all routes have been removed
+
+### Expected issues
+Merging all of our apps together is not without difficulties and we acknowledge that a single app will present problems that our multiple apps do not.
+
+* A Rails update for one large app might be more complex than spread across individual apps.
+* The scaling of a single app might become more costly.
+* It won’t be possible to scale targeted parts of the stack (i.e. one of the existing apps) for expected increases in traffic, instead the whole site will need to be scaled.
+* It’s only possible to test one branch of code in a test environment at once. If there’s only one repo, only one change can be tested at a time.
+* It will be harder to find the pull request history for the retired repositories.
+
+In addition, we will need to address the following issues:
+
+* The test suite will take longer due to more tests.
+    * could potentially parallelise in CI
+    * will need to consider approach for developer machines
+* Need to reconfigure Kubernetes deployment. We should be able to drop the total number of pods.
+    * optimise the number of pods needed in draft and production
+    * currently in production there are
+        * 4 pods for `frontend`
+        * 6 for `government-frontend`
+        * 4 for `collections`
+        * 9 for `finder-frontend`
+        * plus 2 or 3 per app for the others
+    * scaling up the pods for a single app will use more resources
+* Resolve the question around environment testing
+    * rather than testing on integration, we should still be able to test multiple branches using preview apps
+    * would need to get new search working in the preview apps
+
+### After merging
+Once a candidate has been merged, the following steps would be needed.
+
+* Recalibrate Kubernetes deployment
+* Move components from govuk_publishing_components into the app if only being used in the frontend
+    * this information can already be found in our component auditing tools
+    * need to consider which ones are used by publishing applications
+* Begin work to remove static, slimmer and router (to be considered separately from this RFC)
+* Update component auditing
+* Decide how to handle app ownership and maintenance and what that means now we have fewer or only one app
+* Consider renaming `frontend`
+
+## Other options considered
+
+### Convert apps to engines to mount in Frontend
+
+If we wanted to keep the conceptually split functions of the apps but remove the rendering complexity of slimmer/static/router we could recreate the routes and business logic of the existing apps in Rails Engines that would be mounted inside the monolith. This option was rejected on the basis that it would not give us much code-separation advantage over simple namespace discipline within the monolith, and there would still be developer overhead in working with multiple repos.
```

Comment:
We're trialling a new approach in whitehall, where we separate bits of code out into engines, but store them all in the same repo in `lib/engines`. There's only one example of this at the moment, [the content object store](https://github.com/alphagov/whitehall/tree/main/lib/engines/content_object_store).

This would allow us to use rails engines without the overhead of working with multiple repos.

It might be worth considering as a stepping stone, but I don't like it as a long term solution for merging frontends, because it retains the arbitrary split between "apps" (even if they're no longer actually apps).

Reply:
Yeah, if we're going to do this in frontend I'd want to be based on functionality rather than the historical app.